asllib/doc/SubprogramCalls.tex

```diff
@@ -112,150 +91,198 @@ \subsubsection{Formally}
-\hypertarget{def-annotatecallargtyped}{}
+\subsubsection{TypingRule.AnnotateCallActualsTyped \label{sec:TypingRule.AnnotateCallActualsTyped}}
+\hypertarget{def-annotatecallactualstyped}{}
 The function
 \[
   \begin{array}{r}
-  \annotatecallargtyped(
+  \annotatecallactualstyped(
     \overname{\staticenvs}{\tenv} \aslsep
     \overname{\identifier}{\name} \aslsep
-    \overname{(\ty \times\expr)^*}{\callerargstyped} \aslsep
+    \overname{(\ty \times\expr)^*}{\params} \aslsep
+    \overname{(\ty \times\expr)^*}{\texttt{typed\_args}} \aslsep
     \overname{\subprogramtype}{\calltype}
-   ) \aslto \\
-  (\overname{\identifier}{\nameone} \aslsep
-  \overname{\expr^*}{\vargsone} \aslsep
-  \overname{(\identifier\times\expr)^*}{\eqs} \aslsep
+  ) \\ \aslto
+  (\overname{\call}{\vcall} \aslsep
   \overname{\langle \ty \rangle}{\rettyopt})
   \cup \overname{\TTypeError}{\TypeErrorConfig}
 \end{array}
 \]
-is similar to $\annotatecall$, except that the argument expressions are replaced by
-the annotated expressions. That is, pairs consisting of a type and an expression.
+is similar to $\annotatecall$, except that it accepts annotated version of parameter/argument expressions as inputs (that is, pairs consisting of a type and an expression).
 \ProseOtherwiseTypeError
 
+
 \subsubsection{Prose}
 All of the following apply:
 \begin{itemize}
-  \item applying $\unziplist$ to $\callerargtyped$ yields the corresponding lists of types
-        and expressions $\callerargtypes$ and $\vargsone$;
-  \item applying $\subprogramforname$ to match $\name$ and $\callerargtypes$ in $\tenv$
-        yields $(\eqsone, \nameone, \callee)$\ProseOrTypeError;
-  \item checking that $\subprogramtype$ of $\callee$ equals $\calltype$ yields $\True$\ProseOrTypeError;
-  \item checking that the lengths of $\callee.\funcargs$ and $\vargsone$ are the same yields $\True$\ProseOrTypeError;
-  \item applying $\annotateparameterdefining$ to $\callee.\funcargs$,
-        $\callerargstyped$, and \\
-        $\calleeparams$ in $\tenv$ to annotate the implicit parameters
-        yields $\eqsthreep$\ProseOrTypeError;
-  \item define $\eqsthree$ is the concatenation of $\eqsthreep$ and $\eqsone$;
-  \item applying $\checkargstypesat$ to $\calleeargtypes$
-        to check that the actual arguments have correct types with respect to $\callerargtypes$
+  \item applying $\unziplist$ to $\texttt{typed\_args}$ yields the corresponding lists of types
+        and expressions $\argtypes$ and $\args$;
+  \item applying $\subprogramforname$ to match $\name$ and $\argtypes$ in $\tenv$
+        yields \\ $(\namep, \funcsig)$\ProseOrTypeError;
+  \item either the $\subprogramtype$ of $\funcsig$ equals $\calltype$,
+        or the $\subprogramtype$ of $\funcsig$ is $\STGetter$ and $\calltype$ is $\STFunction$\ProseOrTypeError;
+  \item applying $\insertstdlibparam$ to $\funcsig$, $\params$, and $\argtypes$ yields new parameters $\paramsone$;
```

Comment:
But how do you know whether this is a builtin function?
This is a new concept and it's not defined in the reference and it's not part of the abstract syntax. Perhaps there should be an extra field in the abstract syntax and we can add it at the top level to the standard library?

Reply:
In the "Formally" subsubsection for this `\insertstdlibparam`, I attempted to replicate the way evaluation of primitives are handled - in other words, by stating that this transition parametrises the typing/semantics of ASL. Does this seem sufficient, or do you think it needs reworking?

Reply:
I guess it's okay for now.